xc/mem_stats.ml

```diff
@@ -0,0 +1,331 @@
+       The RSS reported by other stat counters could be off depending on the number of threads. *)
+    let smaps_rollup = define_fields ~path:"smaps_rollup" [
+      kib "Rss" "Total amount of resident memory used by %s"
+    ]
+  end
+
+  let find key kv =
+    to_float (KV.get key kv)
+
+  let to_list kv =
+    KV.bindings kv |> List.rev_map snd
+end
+
+module GcStat = struct
+  open SimpleMetrics
+  open Gc
+
+  let ocaml_total =
+    let f = kib "ocaml_total" "Total OCaml memory used by %s" in
+    fun gc control ->
+    gc.heap_words + control.minor_heap_size
+    |> float |> words_to_kib |> f
+
+  let maybe_words name description v =
+    (* quick_stat would return a value of 0, which is not valid *)
+    v |> float |> words_to_kib |> kib ~min:0.001 name description
+
+  let compute (gc, control) =
+    [ ocaml_total gc control
+    ; gc.minor_words +. gc.major_words -. gc.promoted_words
+      |> words_to_kib
+      |> kib_per_s "ocaml_allocation_rate" "Amount of allocations done by OCaml in the given period by %s"
+    ]
+
+  let compute_slow (gc, control) =
+    [
+    (* see https://github.com/ocaml/ocaml/blob/trunk/stdlib/gc.mli#L50-L59, without running a major
+       cycle the live_words may overestimate the actual live words, "live" just means "not currently
+      known to be collectible"*)
+     gc.live_words |> maybe_words "ocaml_maybe_live" "OCaml memory not currently known to be collectable by %s"
+    ; gc.free_words |> maybe_words "ocaml_free" "OCaml memory available to %s"
+    ]
+end
+
+module Derived = struct
+  open SimpleMetrics
+  let memextra_kib =
+    kib "mem_extra" "Total amount of non-OCaml and non-stack memory used by %s"
+
+  let ulimit_stack =
+    let ic = Unix.open_process_in "ulimit -s" in
+    let r = ic |> input_line |> Int64.of_string in
+    close_in_noerr ic ; Int64.to_float r
+
+  let memextra_kib stats (gc_stat, control) =
+    let ocaml_total_kib = GcStat.ocaml_total gc_stat control |> to_float in
+    let vmdata = Proc.find Proc.Fields.vmdata stats in
+    let vmpte = Proc.find Proc.Fields.vmpte stats in
+    let threads = Proc.find Proc.Fields.threads stats in
+      (* Each thread, except the main one will allocate 'ulimit -s' worth of VmData.
+         This won't immediately show up in VmRss until actually used (e.g. by a deep calltree due to recursion) *)
+    vmdata -. vmpte -. ulimit_stack *. (threads -. 1.) -. ocaml_total_kib
+    |> ds_float |> memextra_kib
+end
+
+let observe_stats l =
+  let names = ListLabels.rev_map l ~f:(fun ds -> ds.Ds.ds_name) in
+  let values = ListLabels.rev_map l ~f:(fun ds ->
+    let f = match ds.Ds.ds_value with
+    | Rrd.VT_Int64 i -> Int64.to_float i
+    | Rrd.VT_Float f -> f
+    | Rrd.VT_Unknown -> nan in
+    ds.Ds.ds_pdp_transform_function f |> Printf.sprintf "%.0f")
+  in
+  D.debug "stats header: %s" (String.concat "," names);
+  D.debug "stats values: %s" (String.concat "," values)
+
+let generate_slow_stats =
```

Comment:
The stats are not slow - they are expensive to compute or precise in their result. But if we go with the theme, maybe `slow_stats` and `quick_stats`?